test/e2e/util_gatewayapi_test.go

```diff
@@ -232,6 +233,58 @@ func createGateway(gatewayClass *gatewayapiv1.GatewayClass, name, namespace, dom
 	return gateway, nil
 }
 
+func createGatewayWithListeners(gatewayClass *gatewayapiv1.GatewayClass, name, namespace string, listeners []struct {
+	Name     string
+	Hostname *string
+}) (*gatewayapiv1.Gateway, error) {
+
+	var gatewayListeners []gatewayapiv1.Listener
+	for _, spec := range listeners {
+		var hostname *gatewayapiv1.Hostname = nil
+		if spec.Hostname != nil {
+			h := gatewayapiv1.Hostname(*spec.Hostname)
+			hostname = &h
+		}
+
+		fromNamespace := gatewayapiv1.FromNamespaces(allNamespaces)
+		allowedRoutes := gatewayapiv1.AllowedRoutes{
+			Namespaces: &gatewayapiv1.RouteNamespaces{From: &fromNamespace},
+		}
+
+		l := gatewayapiv1.Listener{
+			Name:          gatewayapiv1.SectionName(spec.Name),
+			Hostname:      hostname,
+			Port:          80,
+			Protocol:      "HTTP",
+			AllowedRoutes: &allowedRoutes,
+		}
+		gatewayListeners = append(gatewayListeners, l)
+	}
+
+	gateway := &gatewayapiv1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: gatewayapiv1.GatewaySpec{
+			GatewayClassName: gatewayapiv1.ObjectName(gatewayClass.Name),
+			Listeners:        gatewayListeners,
+		},
+	}
+
+	if err := kclient.Create(context.TODO(), gateway); err != nil {
+		if kerrors.IsAlreadyExists(err) {
+			name := types.NamespacedName{Namespace: namespace, Name: name}
+			if err = kclient.Get(context.TODO(), name, gateway); err != nil {
+				return nil, fmt.Errorf("failed to get the existing gateway: %v", err.Error())
+			}
+		} else {
+			return nil, fmt.Errorf("failed to create gateway: %v", err.Error())
+		}
```

Comment:
If it already exists, just return an error.  No need to get the existing one.

```suggestion
		if kerrors.IsAlreadyExists(err) {
				return nil, fmt.Errorf("cannot create gateway %s: %v", name, err.Error())
		} else {
			return nil, fmt.Errorf("failed to create gateway %s: %v", name, err.Error())
		}
```